internal/files/manifest.go

```diff
@@ -0,0 +1,183 @@
+package files
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// FileManifest represents a collection of files with their hashes
+type FileManifest struct {
+	Version string      `json:"version"`
+	Files   []FileEntry `json:"files"`
+}
+
+// FileEntry represents a single file in the manifest
+type FileEntry struct {
+	Path string `json:"path"`
+	Hash string `json:"hash"` // MD5 hash to match S3 ETags
+	Size int64  `json:"size"`
+}
+
+// BuildManifest creates a manifest of all files in the directory
+func BuildManifest(rootDir string, ignorePatterns []string) (*FileManifest, error) {
+	manifest := &FileManifest{
+		Version: "1.0",
+		Files:   []FileEntry{},
+	}
+
+	// Create ignore matcher
+	ignoreMatcher := NewIgnoreMatcher(ignorePatterns)
+
+	err := filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Skip directories
+		if info.IsDir() {
+			// Check if directory should be ignored
+			relPath, err := filepath.Rel(rootDir, path)
+			if err != nil {
+				return err
+			}
+			if ignoreMatcher.ShouldIgnore(relPath + "/") {
+				return filepath.SkipDir
+			}
+			return nil
+		}
+
+		// Get relative path
+		relPath, err := filepath.Rel(rootDir, path)
+		if err != nil {
+			return err
+		}
+
+		// Check if file should be ignored
+		if ignoreMatcher.ShouldIgnore(relPath) {
+			return nil
+		}
+
+		// Compute MD5 hash
+		hash, err := computeFileMD5(path)
+		if err != nil {
+			return fmt.Errorf("failed to hash %s: %w", relPath, err)
+		}
+
+		// Add to manifest
+		manifest.Files = append(manifest.Files, FileEntry{
+			Path: relPath,
+			Hash: hash,
+			Size: info.Size(),
+		})
+
+		return nil
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to build manifest: %w", err)
+	}
+
+	return manifest, nil
+}
+
+// computeFileMD5 computes the MD5 hash of a file
+func computeFileMD5(filepath string) (string, error) {
+	file, err := os.Open(filepath)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	hash := md5.New()
+	if _, err := io.Copy(hash, file); err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(hash.Sum(nil)), nil
+}
+
+// CompareManifests compares two manifests and returns what changed
+func CompareManifests(current, previous *FileManifest) (added, modified, deleted []string) {
```

Comment:
Any reason to pass these as pointers? Don't think we need to in this case, we technically need to nilcheck the pointers if we do, just feels like more of a mission than necessary